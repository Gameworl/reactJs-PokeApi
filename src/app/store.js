import { configureStore } from '@reduxjs/toolkit';
import allPokemonSlice from "../features/redux/AllPokemonSlice";
import detailsPokemonSlice from "../features/redux/DetailsPokemonSlice";
import randomPokemonSlice from "../features/redux/RandomPokemonSlice";
import paginationPokemonSlice from "../features/redux/PaginationPokemonSlice";
import favoriePokemon from "../features/redux/FavoriePokemonSlice";
export default configureStore({
  reducer: {
    allPokemon: allPokemonSlice,
    detailsPokemon: detailsPokemonSlice,
    randomPokemon: randomPokemonSlice,
    paginationPokemon: paginationPokemonSlice,
    favoriePokemon: favoriePokemon
  },
});
