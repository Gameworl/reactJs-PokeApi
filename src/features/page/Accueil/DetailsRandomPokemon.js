import React, {useEffect, useState} from "react";
import {ResultRandomPokemon,ResultRandomSpeciePokemon} from "../../redux/RandomPokemonSlice";
import { useSelector} from "react-redux";
import './styles/DetailFocusPokemonStyle.scss';


export const DetailsRandomPokemon = () => {
    const pokemonInfo = useSelector(ResultRandomPokemon)
    const pokemonSpecie = useSelector(ResultRandomSpeciePokemon)
    const [description, setDescription] = useState("");
    const [name, setName] = useState("");
    const [version, setVersion] = useState("");

    useEffect(()=>{
       getResume()
       getName()
    })

    /**
     * recupere le resumé d'un pokemon en Français
     */
    const getResume = () =>{
        pokemonSpecie.flavor_text_entries.forEach(texte => {
            if (texte.language.name === "fr" ){
                if (texte.flavor_text.length !== 0){
                        setDescription(texte.flavor_text)
                        setVersion(texte.version.name)

                    }
                }
        })
    }

    /**
     * Récupère le nom français du pokemon
     */
    const getName = () => {
        pokemonSpecie.names.forEach(name => {
            if(name.language.name === "fr"){
                setName(name.name)
            }
        })

    }



    return(

        <div className="card border-0" style={{width: 18+"rem", height: "fit-content"}}>
            <img
                src={pokemonInfo.sprites.front_default}
                className="card-img-top"
                alt={"image de " + pokemonInfo.name}
            />
            <div className="card-body rounded">
                <h5 className="card-title">{name}</h5>
                <p className="card-text">
                    {description}
                </p>
                <p className="font-italic">Pokedex {version}</p>
            </div>
        </div>

    );
}
