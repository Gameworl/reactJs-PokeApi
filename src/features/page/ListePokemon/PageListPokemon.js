import React, {useEffect,useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {
    StatusPagination,
    queryAllPokemonPagination,
    ResultPaginationPokemon,
    selectedPokemon,
    selectedPage,
    Page
} from "../../redux/PaginationPokemonSlice";
import {
    statusPokemonDetails
} from "../../redux/DetailsPokemonSlice";
import Pagination from "react-js-pagination";
import "./styles/PageListPokemonStyle.scss"
import {Spinner} from "../Spinner/spinner";

import {
    Link,
} from "react-router-dom";
export const PageListPokemon = () => {
    const dispatch = useDispatch()
    const statusList = useSelector(StatusPagination)
    const PokemonList = useSelector(ResultPaginationPokemon)
    const pageNumber = useSelector(Page)
    const [page, setpage] = useState(pageNumber);
    useEffect(()=>{
        if (statusList === "idle"){
            dispatch(
                queryAllPokemonPagination("https://pokeapi.co/api/v2/pokemon?offset=0&limit=10"),
            )
        }
    })

    function handlePageChange(pageNumber) {
        if (page >= 1){
            if (pageNumber > page){
                dispatch(
                    queryAllPokemonPagination(PokemonList.next),
                )
            }else {
                dispatch(
                    queryAllPokemonPagination(PokemonList.previous),
                )
            }
            dispatch(
                selectedPage({
                    page : pageNumber
                })
            )
            setpage(pageNumber)
        }
    }
    const handleClick = (pokemon) => {
        dispatch(
            statusPokemonDetails({
                StatusDetails: "idle",
                StatusDetailsSpecie: "idle"
            }),
        )
        dispatch(
            selectedPokemon({
                selected : pokemon.url
            })
        )
    }


    return(
        <div className="d-flex ">
            {statusList === "succeeded"?
            <div className="row">
                {statusList === "succeeded" ? PokemonList.results.map((pokemon) => (
                    <div key={pokemon.name} className=" col-md-2 m-4 h-25 card align-items-center transparentB" >
                        <img className="pokeIcon " src={"https://img.pokemondb.net/sprites/home/normal/"+ pokemon.name+".png"} alt={"image de " + pokemon.name} />
                        <h5 className="card-title justify-content-center">{pokemon.name}</h5>
                        <Link to={"/details"} className="btn border-0 transparent" onClick={() =>handleClick(pokemon)}>
                            <img className="logo" src={"/assets/pokeIcon.jpg"} alt={"logo"}/>
                            DÃ©tails
                            <img className="logo" src={"/assets/pokeIcon.jpg"} alt={"logo"}/>
                        </Link>
                    </div>
                )) : ""}
                <div className="m-auto h-auto">
                    <Pagination
                        itemClass="page-item "
                        linkClass="page-link"
                        hideFirstLastPages
                        prevPageText='prev'
                        nextPageText='next'
                        activePage={page}
                        itemsCountPerPage={10}
                        totalItemsCount={1118}
                        pageRangeDisplayed={3}
                        onChange={handlePageChange.bind(this)}
                    />
                </div>
            </div>: <Spinner/>}

        </div>

    )
}


