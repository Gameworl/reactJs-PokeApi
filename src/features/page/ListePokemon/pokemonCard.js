import React from "react";
import {Link} from "react-router-dom";
import {statusPokemonDetails} from "../../redux/DetailsPokemonSlice";
import {selectedPokemon} from "../../redux/PaginationPokemonSlice";
import {useDispatch} from "react-redux";
export const PokemonCard = (props) => {
    const dispatch = useDispatch()
    const image = (urlImage) => {
        let image = urlImage.split('/');
        let imagePokemon = image[6];
        return imagePokemon
    };
    const handleClick = (pokemon) => {
        dispatch(
            statusPokemonDetails({
                StatusDetails: "idle",
                StatusDetailsSpecie: "idle"
            }),
        )
        dispatch(
            selectedPokemon({
                selected : pokemon.url
            })
        )
    }

    return(
        <div key={image(props.pokemon.url)} className=" col-md-2 m-4 h-25 card align-items-center transparentB" >
            <img className="pokeIcon " src={"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/"+image(props.pokemon.url)+".png"} alt={"image de " + props.pokemon.name} />
            <h5 className="card-title justify-content-center">{props.pokemon.name}</h5>
            <Link to={"/details"} className="btn border-0 transparent" onClick={() =>handleClick(props.pokemon)}>
                <img className="logo" src={"/assets/pokeIcon.jpg"} alt={"logo"}/>
                DÃ©tails
                <img className="logo" src={"/assets/pokeIcon.jpg"} alt={"logo"}/>
            </Link>
        </div>
        );
}
