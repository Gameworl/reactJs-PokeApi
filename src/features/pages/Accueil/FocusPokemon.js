import React, {useEffect} from "react";
import {
    queryRandomPokemon,
    queryRandomSpeciePokemon,
    ResultRandomPokemon,
    statusPokemonRandom,
    StatusRandom,
    StatusRandomSpecie
} from "../../redux/RandomPokemonSlice";
import {
    ResultsPokemonList,
    StatusList,
    queryAllPokemon
} from "../../redux/AllPokemonSlice";
import {useDispatch,useSelector} from "react-redux";
import {DetailsRandomPokemon} from "./DetailsRandomPokemon";
import './styles/FocusOnePokemonStyle.scss';
import {Spinner} from "../Spinner/spinner";

/**
 * Page contenant un pokemon choisis aléatoirement
 * @returns {*}
 * @constructor
 */
export const FocusPokemon = () => {
    const dispatch = useDispatch()
    const pokemonUrl = useSelector(ResultsPokemonList)
    const statusList = useSelector(StatusList)
    const statusOne = useSelector(StatusRandom)
    const statusSpecie = useSelector(StatusRandomSpecie)
    const pokemonInfo = useSelector(ResultRandomPokemon)

    useEffect(() => {
        if (statusList === "idle") {
            dispatch(
                queryAllPokemon(),
            )
        }
        if (statusList === "succeeded" && statusOne === "idle"){
            dispatch(
                queryRandomPokemon(pokemonUrl[Math.floor(Math.random() * 1118)].url),
            )
        }
        if (statusOne === "succeeded" && statusSpecie === "idle"){
            dispatch(
                queryRandomSpeciePokemon(pokemonInfo.species.url)
            )
        }
    })
    /**
     * Refresh le status des requêtes qui selectionne le pokemon aléatoire
     */
    const refreshPage = () => {
        dispatch(
            statusPokemonRandom({
                statusRandomPokemonQuery: "idle",
                statusRandomSpeciePokemonQuery: "idle"
            }),
        )
    }
    return(
        <div className="col d-flex justify-content-center background align-items-center">
            {statusSpecie === "succeeded"? <button onClick={refreshPage} className="float-right transparentB border-0" style={{ cursor: "pointer"}}>
                <img src={"/assets/refresh.png"} style={{height : 2+"rem", width: 2+"rem"}} alt={"refreshIcon"}/>
            </button>:""}
            {statusSpecie === "succeeded"? <DetailsRandomPokemon/> : <Spinner/>}
        </div>
    );

}
