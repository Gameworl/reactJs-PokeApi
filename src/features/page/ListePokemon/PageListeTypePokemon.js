import React, {useEffect,useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {
    StatusListeTypePokemon,
    ResultListeTypePokemon,
    queryListeTypePokemon, StatusTypePokemon, ResultTypePokemon,queryTypePokemon
} from "../../redux/PaginationPokemonSlice";
import Pagination from "react-js-pagination";
import "./styles/PageListPokemonStyle.scss"
import {Spinner} from "../Spinner/spinner";

import {PokemonCard} from "./pokemonCard";
import {SelectionTypeBar} from "./SelectionTypeBar";
export const PageListeTypePokemon = () => {
    const dispatch = useDispatch()
    const statusList = useSelector(StatusListeTypePokemon)
    const statusType = useSelector(StatusTypePokemon)
    const TypeList = useSelector(ResultTypePokemon)
    const PokemonList = useSelector(ResultListeTypePokemon)
    const [page, setpage] = useState(1);

    useEffect(()=>{
        if (statusType === "idle"){
            dispatch(
                queryTypePokemon(),
            )
        }
    })

    const handlePageChange = (pageNumber) => {
        if (page >= 1){
            setpage(pageNumber)
        }
    }


    const maRecherche= (texte) => {
        if (texte.lenght !== 0){
            dispatch(
                queryListeTypePokemon(texte),
            )
        }
    }

    return(
        <div className="d-flex ">
            <div className="col">
                { statusType === "succeeded" ?<div className="row justify-content-center m-3" style={{ height: "fit-content"}}>
                    <SelectionTypeBar types={TypeList} recherche={maRecherche}/>
                </div>: ""}

            {statusList === "succeeded"?
                <div className="row"  style={{ height: "fit-content"}}>
                    {statusList === "succeeded" ? PokemonList[page-1].map((pokemon) => (
                        <PokemonCard pokemon={pokemon.pokemon}/>
                    )) : ""}
                    <div className="m-auto h-auto">
                        <Pagination
                            itemClass="page-item "
                            linkClass="page-link"
                            hideFirstLastPages
                            prevPageText='prev'
                            nextPageText='next'
                            activePage={page}
                            itemsCountPerPage={10}
                            totalItemsCount={PokemonList.length*10}
                            pageRangeDisplayed={3}
                            onChange={handlePageChange.bind(this)}
                        />
                    </div>
                </div>: <Spinner/>}
            </div>
        </div>

    )
}


