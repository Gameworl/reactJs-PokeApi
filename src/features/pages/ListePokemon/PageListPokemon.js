import React, {useEffect,useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {
    StatusPagination,
    queryAllPokemonPagination,
    ResultPaginationPokemon,
    selectedPage,
    Page
} from "../../redux/PaginationPokemonSlice";
import Pagination from "react-js-pagination";
import "./styles/PageListPokemonStyle.scss"
import {Spinner} from "../Spinner/spinner";
import {PokemonCard} from "./pokemonCard";

/**
 * Page contenant la liste de tout les pokemons
 * @returns {*}
 * @constructor
 */
export const PageListPokemon = () => {
    const dispatch = useDispatch()
    const statusList = useSelector(StatusPagination)
    const PokemonList = useSelector(ResultPaginationPokemon)
    const pageNumber = useSelector(Page)
    const [page, setpage] = useState(pageNumber);
    useEffect(()=>{
        if (statusList === "idle"){
            dispatch(
                queryAllPokemonPagination("https://pokeapi.co/api/v2/pokemon?offset=0&limit=10"),
            )
        }
    })
    /**
     * fait des appels en fonction du bouton cliquÃ© sur la pagination
     * @param pageNumber
     */
    const handlePageChange = (pageNumber) =>{
        if (page >= 1){
            if (pageNumber > page){
                dispatch(
                    queryAllPokemonPagination(PokemonList.next),
                )
            }else {
                dispatch(
                    queryAllPokemonPagination(PokemonList.previous),
                )
            }
            dispatch(
                selectedPage({
                    page : pageNumber
                })
            )
            setpage(pageNumber)
        }
    }



    return(
        <div className="d-flex ">
            {statusList === "succeeded"?
            <div  className="row">
                {statusList === "succeeded" ? PokemonList.results.map((pokemon) => (
                    <PokemonCard pokemon={pokemon}/>
                )) : ""}
                <div className="m-auto h-auto">
                    <Pagination
                        itemClass="page-item "
                        linkClass="page-link"
                        hideFirstLastPages
                        prevPageText='prev'
                        nextPageText='next'
                        activePage={page}
                        itemsCountPerPage={10}
                        totalItemsCount={1118}
                        pageRangeDisplayed={3}
                        onChange={handlePageChange.bind(this)}
                    />
                </div>
            </div>: <Spinner/>}
        </div>
    )
}


