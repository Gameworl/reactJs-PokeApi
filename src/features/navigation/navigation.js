import React from "react";
import {
    BrowserRouter as Router,
    Switch,
    Route,
    Link,
} from "react-router-dom";
import {FocusPokemon} from "../page/Accueil/FocusPokemon";
import "./styles/navigationStyle.scss"
import {PageListPokemon} from "../page/ListePokemon/PageListPokemon";
import {Pokemon} from "../page/DetailsPokemon/Pokemon";
import {Mespokemon} from "../page/Favorie/mesProkemon";
import {PageListeTypePokemon} from "../page/ListePokemon/PageListeTypePokemon";

export const Navigation = () => {
    return(
        <Router >
            <div>
                <nav className="navbar navbar-expand-lg navbar-light bg-light">
                        <Link className="navbar-brand" to="/">
                            <img src={"/assets/logo.png"} width="30" height="30"
                                 className="d-inline-block align-top" alt=""/>
                        </Link>
                    <div className="collapse navbar-collapse" id="navbarSupportedContent">
                        <ul className="navbar-nav mr-auto">
                            <li className="nav-item">
                                <Link  className={"nav-link"} to="/">Accueil</Link>
                            </li>
                            <li className="nav-item">
                                <Link  className={"nav-link"} to="/PokeList">PokeList</Link>
                            </li>
                            <li className="nav-item">
                                <Link  className={"nav-link"} to="/PokeListType">TypeList</Link>
                            </li>
                            <li className="nav-item">
                                <Link  className={"nav-link"} to="/MesPokemons">Mes Pokemons</Link>
                            </li>

                        </ul>
                    </div>
                </nav>

                {/* A <Switch> looks through its children <Route>s and
                  renders the first one that matches the current URL. */}
                <Switch>
                    <Route exact path="/PokeList">
                        <div className={"backgroundList"}>
                            <PageListPokemon/>

                        </div>
                    </Route>
                    <Route exact path="/">
                        <div className={"backgroundAccueil"} style={{backgroundImage: "/background/background.jpg"}}>
                            <FocusPokemon/>
                        </div>
                    </Route>
                    <Route exact path="/details">
                        <div className={"backgroundAccueil"}>
                            <Pokemon/>
                        </div>
                    </Route>
                    <Route exact path="/PokeListType">
                        <div className={"backgroundList"}>
                            <PageListeTypePokemon/>
                        </div>
                    </Route>
                    <Route exact path="/MesPokemons">
                        <div className={"backgroundList"}>
                            <Mespokemon/>
                        </div>
                    </Route>
                    <Route path="*">
                        <div className={"background404"}>
                        </div>
                    </Route>
                </Switch>
            </div>
        </Router>
    )

}
