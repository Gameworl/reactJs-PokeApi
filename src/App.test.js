import React from 'react';
import { render,screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import store from './app/store';
import {Navigation} from "./features/navigation/navigation";
import {FocusPokemon} from "./features/page/Accueil/FocusPokemon";
import {PageListeTypePokemon} from "./features/page/ListePokemon/PageListeTypePokemon";

describe('Navigation', ()=>{
    test('Test d\'un mots sur une page', () => {
        const { getByText } = render(
            <Provider store={store}>
                <Navigation/>
            </Provider>
        );
        expect(getByText(/Accueil/i)).toBeInTheDocument();
    });

    test('Test si on a une navBar', () => {
        render(
            <Provider store={store}>
                <Navigation/>
            </Provider>
        );
        expect(screen.getByRole('navigation')).toBeInTheDocument();
    });

    test('Test nb lien de redirection', () => {
        render(
            <Provider store={store}>
                <Navigation/>
            </Provider>
        );
        const listLink = screen.getAllByRole('link');
        expect(listLink.length).toEqual(5);
    });
});

describe('Page Liste type Pokemon ', ()=>{

    test('Test que aucun type est charger', async () => {
        render(
            <Provider store={store}>
                <PageListeTypePokemon/>
            </Provider>
        );

        expect(screen.getByRole('img')).toBeInTheDocument();
    });

    test('Test Un type est selectionner', async () => {
        render(
            <Provider store={store}>
                <PageListeTypePokemon/>
            </Provider>
        );

        expect(screen.queryByText(/Détails/)).toBeNull();
        await userEvent.type(screen.getByRole('option'), 'dragon');
        expect(screen.findByText(/Détails/));
    });

});

