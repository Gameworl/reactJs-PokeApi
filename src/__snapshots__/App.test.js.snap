// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Favorie Pokemon test retourne la state initial 1`] = `
Object {
  "resultsFavorie": Array [],
}
`;

exports[`Test Redux test AllPokemonSlice retourne la state initial 1`] = `
Object {
  "resultsAllPokemon": Array [],
  "statusAllPokemonQuery": "idle",
}
`;

exports[`Test Redux test DetailsPokemonSlice retourne la state initial 1`] = `
Object {
  "resultDetailsPokemon": Array [],
  "resultDetailsSpeciePokemon": Array [],
  "statusDetailsPokemonQuery": "idle",
  "statusDetailsSpeciePokemonQuery": "idle",
}
`;

exports[`Test Redux test FavoriePokemonSlice retourne la state initial 1`] = `
Object {
  "resultsFavorie": Array [],
}
`;

exports[`Test Redux test PaginationPokemonSlice retourne la state initial 1`] = `
Object {
  "listeType": Array [],
  "page": 1,
  "resultPaginationPokemon": Array [],
  "resultlisteType": Array [],
  "selected": "",
  "statusPaginationPokemonQuery": "idle",
  "statuslisteType": "idle",
  "statuslisteTypePokemonQuery": "idle",
}
`;

exports[`Test Redux test RandomPokemonSlice retourne la state initial 1`] = `
Object {
  "resultRandomPokemon": Array [],
  "resultRandomSpeciePokemon": Array [],
  "statusRandomPokemonQuery": "idle",
  "statusRandomSpeciePokemonQuery": "idle",
}
`;
